Homework1. Created Storage and covered it with tests

marl0vv opened the pull request.

marl0vv committed:
Create project

Added new project
Added maven dependencies
Added file with input data

marl0vv committed:
Added Storage and read from file

Added:
Record Entry that contains information for one element from json
Class Storage that contains entries
Class Input that reads entries from .json and saves them into storage

marl0vv committed:
Added Service

Added class Service that works with user

marl0vv committed:
Update Main to run program

marl0vv committed:
Added tests

RavenVSS commented on src/main/java/com/marl0vv/study/Input.java:
```diff
@@ -0,0 +1,53 @@
+public class Input {
+    private static int generateUniqueID(Map<Integer, Entry> storage, int startingId) {
+        int newId = startingId;
+        while (storage.containsKey(newId)) {
```
А если использовать `storage.size()` вместо `startingId`?  Чтобы от лишнего аргумента избавится

marl0vv replied:
Сначала мне показалось, что это однозначно хорошая идея, потому что у меня startingId по сути и представляет размер storage. Только я предполагал, что среди данных в json может быть id, а может и не быть id. Однако в ситуации, когда в json есть id и он идёт не на следующую позицию, а на что-то дальше лежащее, storage увеличится, но на следующую позицию ничего не добавится. Мы получим следующую ситуацию, как на скриншоте, когда запись "Southern Air" идёт на пятое место, так как у неё указан id, но следующая запись "Unknown pleasures" пойдёт на третье место, хотя я ожидал второе. Если нам не важен порядок id, то ладно, но мне как-то кажется, что вот так пропускать доступный id нехорошо. 
![Screenshot 2023-10-28 183518](https://github.com/marl0vv/whitesoft-2023-backend-storage/assets/73769634/8e5a4ab4-a41a-43c9-9015-c37fdcdb645b)
![Screenshot 2023-10-28 183853](https://github.com/marl0vv/whitesoft-2023-backend-storage/assets/73769634/1e6b75a3-e764-47b9-9b44-1562958a9121)

RavenVSS replied:
То что пропускаешь id это не критично. Есть мощные проекты где и по 5000 id пропускается =)
Главное чтобы id был всегда уникальным.

В идеале сделать одну глобальную переменную, которая будет выдавать nextId и при каждом новом объекте её увеличивать. Будут пропуски, но это и лучше, т.к. если удалили объект с id = 1, потом за место него добавили другой, получается уже id = 1 будет вести на другой объект

RavenVSS commented on src/main/java/com/marl0vv/study/Input.java:
```diff
@@ -0,0 +1,53 @@
+        for (Entry entry : entries) {
+            if (entry.id() == 0 || storage.containsKey(entry.id())) {
+                int newId = generateUniqueID(storage, nextId);
+                Entry updatedEntry = entry.changeId(newId);
```
Тут можно просто `entry.setId(newId);` и далее в `put` закидывать просто `entry`

marl0vv replied:
Исправлено. Надеюсь, что правильно понял, потому что у меня Entry был Record и я не просто менял Id, а создавал новый объект с новым Id. Поэтому я переписал весь Entry, чтобы он был классом, добавил сеттеры и теперь передаю в put только изначальный Entry с изменённым id

RavenVSS commented on src/main/java/com/marl0vv/study/Main.java:
```diff
@@ -0,0 +1,9 @@
+
+public class Main {
+    public static void main(String[] args) {
+        Storage storage = new Storage("./src/main/resources/inputData.json");
```
По заданию 
```
при запуске приложение принимает в качестве параметра путь до файла, 
из которого необходимо считать данные (формат по желанию)
```
Нужно этот путь либо через `args` получать, либо через `System.in`, если есть третий вариант, можно и его :). В тестах можно путь в коде указывать

marl0vv replied:
Исправлено.

RavenVSS commented on src/main/java/com/marl0vv/study/Service.java:
```diff
@@ -0,0 +1,77 @@
+            int optionChoice;
+            if (sc.hasNextInt()){
+                optionChoice = sc.nextInt();
+                sc.nextLine(); // Consume the newline character
```
nit-pick: Можно всё в одну строчку сделать вот так:
```
Integer.parseInt(sc.nextLine());
```

marl0vv replied:
Удобно. Добавил в код.

RavenVSS commented on src/main/java/com/marl0vv/study/Storage.java:
```diff
@@ -0,0 +1,25 @@
+public class Storage {
+    private Map<Integer, Entry> storage;
+    public Storage(String filePath){
+        this.storage = Input.readFromFile("./src/main/resources/inputData.json");
```
Параметр `filePath` нигде не используется, прокинь в `Input.readFromFile`

marl0vv replied:
Исправлено.

RavenVSS commented on src/test/java/com/marl0vv/study/InputTest.java:
```diff
@@ -0,0 +1,29 @@
+
+        Entry expectedEntry = new Entry(1, "The Shape of Punk To Come", "Hardcore punk", "https://en.wikipedia.org/wiki/The_Shape_of_Punk_to_Come");
+        Entry actualEntry = storage.get(1);
+        assertNotNull(actualEntry);
```
Лишняя проверка, внизу assertEquals проверит на null

marl0vv replied:
Убрал `assertNotNull`

RavenVSS commented on src/test/java/com/marl0vv/study/ServiceTest.java:
```diff
@@ -0,0 +1,57 @@
+
+public class ServiceTest {
+
+    @Mock
```
Нигде не используется, можно убрать
Если хочешь проверить вызовы, то сделай в методе setUp так:
```
Mockito.spy(new Storage("./src/test/resources/inputData.json"));
```
Потом можешь проверять вызовы методов через `Mockito.verify`

Но в твоём случае Storage никуда не пробрасывается, поэтому можно вообще без Mock обойтись

marl0vv replied:
Убрал

RavenVSS commented on src/test/java/com/marl0vv/study/ServiceTest.java:
```diff
@@ -0,0 +1,57 @@
+
+    @BeforeEach
+    public void setUp(){
+        storage = new Storage("./src/test/resources/inputData.json");
```
Метод можно убрать, а `new Storage("./src/test/resources/inputData.json");` сразу в поле класса прокинуть
```
private Storage storage = new Storage("./src/test/resources/inputData.json");
```
Вообще аннотация `@BeforeEach` будет запускать метод после каждого теста, в твоём случае нужно вешать `@Before`, тогда метод только один раз выполнится перед первым тестом. Но как я написал выше, можно без этого метода обойтись

marl0vv replied:
Перекинул `new Storage("./src/test/resources/inputData.json");` в поле класса. Метод убрал.

marl0vv committed:
bugfix

Put generated newId in storage instead of nextId

marl0vv committed:
Made entry from Record into Class

In order not to create new Entry objects when changing id, I changed Entry from Record to Class.
Included lombok in Maven to avoid writing getters myself
Added function to print Entry because I can't print class attributes like Records
Changed all places where getters were needed

marl0vv committed:
fix path to inputData

Path to inputData is now read from console arguments
Constructor in Storage now uses filePath variable

marl0vv committed:
Used Integer.parseInt

marl0vv committed:
Tests fixes
